lib/shoryuken/processor.rb

```diff
@@ -1,15 +1,79 @@
-require 'json'
-
 module Shoryuken
   class Processor
-    include Celluloid
     include Util
 
-    def initialize(manager)
-      @manager = manager
+    BATCH_LIMIT = 10
+
+    def initialize(fetcher, polling_strategy)
+      @fetcher = fetcher
+      @polling_strategy = polling_strategy
+      # what is the worst thing that could happen because it's not atomic?
+      # is it possible that it evaluates to true even if we never set it to true?
+      @done = false
+    end
+
+    def start
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      # can it throw exception?
+      # what should we do if it dies?
+      @thread ||= spawn_thread(&method(:loop))
+    end
+
+    def terminate(wait=false)
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.value if wait
+    end
+
+    def kill
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.raise ::Shoryuken::Shutdown
+    end
```

Comment:
@mariokostelac 

> We actually spend 980 ms waiting for next dispatch.
  
How long it would it take to perform 8 requests to SQS to fetch a message per processor?

But anyway, I mentioned in another comment, we can remove the `heartbeat` and call `dispatch` once a processor is done. So we processor whenever a processor is ready.

[`TimeTask`](http://www.rubydoc.info/github/ruby-concurrency/concurrent-ruby/Concurrent/TimerTask) says that the delay is in seconds, but maybe it supports something like this `0.5`.